de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/hibernate/PageDAO.java

```diff
@@ -19,110 +19,56 @@
-
-    private final SessionFactory sessionFactory;
-
-    private Wikipedia wiki;
-
     public PageDAO(Wikipedia pWiki) {
-        this.wiki = pWiki;
-        sessionFactory = getSessionFactory();
-    }
-
-    protected SessionFactory getSessionFactory() {
-        try {
-            return WikiHibernateUtil.getSessionFactory(wiki.getDatabaseConfiguration());
-        } catch (Exception e) {
-            logger.error("Could not locate SessionFactory in JNDI", e);
-            throw new IllegalStateException("Could not locate SessionFactory in JNDI");
-        }
+        super(pWiki, "de.tudarmstadt.ukp.wikipedia.api.hibernate.Page");
     }
 
+    @Override
     public void persist(Page transientInstance) {
         logger.debug("persisting Page instance");
-        try {
-            sessionFactory.getCurrentSession().persist(transientInstance);
-            logger.debug("persist successful");
-        } catch (RuntimeException re) {
-            logger.error("persist failed", re);
-            throw re;
-        }
+        super.persist(transientInstance);
     }
 
+    @Override
     public void attachDirty(Page instance) {
         logger.debug("attaching dirty Page instance");
-        try {
-            sessionFactory.getCurrentSession().saveOrUpdate(instance);
-            logger.debug("attach successful");
-        } catch (RuntimeException re) {
-            logger.error("attach failed", re);
-            throw re;
-        }
+        super.attachDirty(instance);
     }
 
+    @Override
     public void attachClean(Page instance) {
         logger.debug("attaching clean Page instance");
-        try {
-//            sessionFactory.getCurrentSession().lock(instance, LockMode.NONE);
-            sessionFactory.getCurrentSession().buildLockRequest(LockOptions.NONE).lock(instance);
-            logger.debug("attach successful");
-        } catch (RuntimeException re) {
-            logger.error("attach failed", re);
-            throw re;
-        }
+        super.attachClean(instance);
     }
 
+    @Override
     public void delete(Page persistentInstance) {
         logger.debug("deleting Page instance");
-        try {
-            sessionFactory.getCurrentSession().delete(persistentInstance);
-            logger.debug("delete successful");
-        } catch (RuntimeException re) {
-            logger.error("delete failed", re);
-            throw re;
-        }
+        super.delete(persistentInstance);
     }
 
+    @Override
     public Page merge(Page detachedInstance) {
         logger.debug("merging Page instance");
-        try {
-            Page result = (Page) sessionFactory.getCurrentSession().merge(detachedInstance);
-            logger.debug("merge successful");
-            return result;
-        } catch (RuntimeException re) {
-            logger.error("merge failed", re);
-            throw re;
-        }
+        return super.merge(detachedInstance);
```

Comment:
How about adding a logging message to the `super.merge` which logs something like `"Failed merging " + daoClass + " instance"`? Overwriting this in every DAO subclass seems very redundant.

Reply:
The redundancy is already gone by introducing `GenericDAO` itself. I will improve the logging output `GenericDAO` for this case this way: 

```
    public T merge(T detachedInstance) {
        try {
            T result = (T) getSession().merge(detachedInstance);
            logger.trace("merge successful");
            return result;
        } catch (RuntimeException re) {
            logger.error("Failed merging " + entityClass + " instance", re);
            throw re;
        }
    }
```

Thus, we always find the actual class name which had merge failures. I will this pattern to other methods in `GenericDAO` as well.

Reply:
I will add `logger.error("Failed merging " + entityClass + " instance", re);` in GenericDAO. 
Similar with the other methods.